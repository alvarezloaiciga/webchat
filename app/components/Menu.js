// @flow
import React from 'react';
import {css} from 'emotion';
import styled from 'react-emotion';

type MenuItemType = {
  onClick: () => void,
  label: string,
  title?: string,
  id: string,
  icon?: {
    name: string,
    style?: Object,
  },
  style?: Object,
};

export type MenuProps = {
  items: Array<MenuItemType>,
  className?: string,
  containerStyle?: Object,
};

const hoverBackground = (
  normalBg: string,
  hoverBg: string,
  activeBg: string,
  selectedBg?: string,
  selected?: boolean,
) => css`
  transition: 0.15s ease-in-out all;
  background: ${selected ? selectedBg : normalBg};

  &:hover {
    background: ${selected ? selectedBg : hoverBg};
  }

  &:active {
    background: ${selected ? selectedBg : activeBg};
  }
`;

const MenuContainer = styled.div`
  @keyframes dropDownOpen {
    0% {
      opacity: 0;
      transform: scale(0.9);
    }
    50% {
      opacity: 0.7;
      transform: scale(1.1);
    }
    100% {
      opacity: 1;
      transform: scale(1);
    }
  }

  flex: 0 0 auto;
  z-index: 10;
  background: #fff;
  border: 1px solid #e2e3e5;
  color: darkgery;
  font-size: 15px;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.2);
  animation: dropDownOpen 0.3s 1;
  cursor: pointer;
  display: flex;
  flex-direction: column;
`;

const MenuItem = styled.div`
  transition: 0.15s ease-in-out all;
  padding: 5px;
  flex: 0 0 auto;
  display: flex;

  ${hoverBackground('transparent', 'rgba(0,0,0,0.15)', 'rgba(0,0,0,0.25)')};
`;

const Icon = styled.i`margin-right: 5px;`;

const Text = styled.div``;

const Menu = (props: MenuProps) => {
  return (
    <MenuContainer className={props.className || ''} style={props.containerStyle}>
      {props.items.map(i => (
        <MenuItem onClick={i.onClick} key={i.id} title={i.title || i.label} style={i.style}>
          {i.icon && <Icon style={i.icon.style} className={`fa fa-fw fa-${i.icon.name}`} />}
          <Text>{i.label}</Text>
        </MenuItem>
      ))}
    </MenuContainer>
  );
};

export default Menu;
