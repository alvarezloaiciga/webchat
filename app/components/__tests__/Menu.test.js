// @flow
import React from 'react';
import type {MenuProps} from '../Menu';
import Menu from '../Menu';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';

const mockMenuItem = (overrides?: Object = {}) => {
  const obj = {
    onClick: jest.fn(),
    label: 'label',
    title: 'title',
    id: 'id',
    disabled: false,
    icon: {
      name: 'iconName',
      style: {},
    },
    style: {},
  };

  return Object.assign({}, obj, overrides);
};

describe('Menu component', () => {
  let wrapper: ShallowWrapper;
  let testProps: MenuProps;
  let render: () => void;

  beforeEach(() => {
    testProps = {
      items: [mockMenuItem()],
      className: 'PICKLE_RIIIIIIIIIICCCCCCKKKKKK',
      containerStyle: {},
    };
    render = () => {
      wrapper = shallow(<Menu {...testProps} />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('minimal props', () => {
      it('renders', () => {
        testProps = {
          items: [],
        };

        render();
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("when title isn't available", () => {
    it('uses label', () => {
      testProps.items = [mockMenuItem({title: undefined})];
      render();
      expect(wrapper.find('[data-test="menuItem"]').prop('title')).toBe(testProps.items[0].label);
    });
  });
});
