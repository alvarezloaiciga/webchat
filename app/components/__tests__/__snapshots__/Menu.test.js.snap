// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component rendering minimal props renders 1`] = `
<Styled(div)
  className=""
/>
`;

exports[`Menu component rendering renders 1`] = `
<Styled(div)
  className="PICKLE_RIIIIIIIIIICCCCCCKKKKKK"
  style={Object {}}
>
  <Styled(div)
    data-test="menuItem"
    disabled={false}
    onClick={[Function]}
    style={Object {}}
    title="title"
  >
    <Styled(i)
      className="fa fa-fw fa-iconName"
    />
    <Styled(div)>
      label
    </Styled(div)>
  </Styled(div)>
</Styled(div)>
`;
